lib/src/outputs/advanced_file_output.dart

```diff
@@ -0,0 +1,147 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import '../log_level.dart';
+import '../log_output.dart';
+import '../output_event.dart';
+
+extension _NumExt on num {
+  String get twoDigits => toString().padLeft(2, '0');
+  String get threeDigits => toString().padLeft(3, '0');
+}
+
+/// Writes the log output to a file.
+class AdvancedFileOutput extends LogOutput {
+  AdvancedFileOutput({
+    this.directory,
+    this.file,
+    this.overrideExisting = false,
+    this.encoding = utf8,
+    List<Level>? writeImmediately,
+    this.maxDelay = const Duration(seconds: 2),
+    this.maxBufferSize = 2000,
+    this.maxLogFileSizeMB = 1,
+  })  : writeImmediately = writeImmediately ??
+            [
+              Level.error,
+              Level.fatal,
+              Level.warning,
+              // ignore: deprecated_member_use_from_same_package
+              Level.wtf,
+            ],
+        assert(
+          (file != null ? 1 : 0) + (directory != null ? 1 : 0) == 1,
+          'Either file or directory must be set',
+        );
+
+  final File? file;
+  final Directory? directory;
+
+  final bool overrideExisting;
+  final Encoding encoding;
+
+  final List<Level> writeImmediately;
+  final Duration maxDelay;
+  final int maxLogFileSizeMB;
+  final int maxBufferSize;
+
+  IOSink? _sink;
+  File? _targetFile;
+  Timer? _bufferWriteTimer;
+  Timer? _targetFileUpdater;
+
+  final List<OutputEvent> _buffer = [];
+
+  bool get dynamicFilesMode => directory != null;
+  File? get targetFile => _targetFile;
+
+  @override
+  Future<void> init() async {
+    if (dynamicFilesMode) {
+      //we use sync directory check to avoid losing
+      //potential initial boot logs in early crash scenarios
+      if (!directory!.existsSync()) {
+        directory!.createSync(recursive: true);
+      }
+
+      _targetFileUpdater = Timer.periodic(
+        const Duration(minutes: 1),
+        (_) => _updateTargetFile(),
+      );
+    }
+
+    _bufferWriteTimer = Timer.periodic(maxDelay, (_) => _writeOutBuffer());
+    await _updateTargetFile(); //run first setup without waiting for timer tick
+  }
+
+  @override
+  void output(OutputEvent event) {
+    _buffer.add(event);
+    // If event level is present in writeImmediately, write out the buffer
+    // along with any other possible elements that accumulated in it since
+    // the last timer tick
+    // Also write out if buffer is overfilled
+    if (_buffer.length > maxBufferSize ||
+        writeImmediately.contains(event.level)) {
+      _writeOutBuffer();
+    }
+  }
+
+  void _writeOutBuffer() {
+    if (_sink == null) return; //wait until _sink becomes available
+    for (final event in _buffer) {
+      _sink?.writeAll(event.lines, Platform.lineTerminator);
+      _sink?.writeln();
+    }
+    _buffer.clear();
+  }
```

Comment:
Not sure... Especially for frequent writes (like a stream of errors) this will result in additional overhead of size checking on every write operation. I'd rather stick with a timer solution and allow the files to be slightly bigger than the size limit, rather than doing this on every write operation.

Reply:
The size check is a good point! I mean, we could still try to track the size by ourselves ;)
Anyway, your approach seems sensible to me then!